src/Lib.P2P/SecureCommunication/Psk1Stream.cs

```diff
@@ -0,0 +1,203 @@
+                    _writeCipher = new XSalsa20Engine();
+                    _writeCipher.Init(true, new ParametersWithIV(new KeyParameter(_key.Value), nonce));
+                }
+
+                return _writeCipher;
+            }
+        }
+
+        private IStreamCipher ReadCipher
+        {
+            get
+            {
+                if (_readCipher == null)
+                {
+                    // Get the nonce from the remote.
+                    var nonce = new byte[NonceByteLength];
+                    for (int i = 0, n; i < NonceByteLength; i += n)
+                    {
+                        n = _stream.Read(nonce, i, NonceByteLength - i);
+                        if (n < 1)
+                            throw new EndOfStreamException();
+                    }
+
+                    // Create the cipher
+                    _readCipher = new XSalsa20Engine();
+                    _readCipher.Init(false, new ParametersWithIV(new KeyParameter(_key.Value), nonce));
+                }
+
+                return _readCipher;
+            }
+        }
+
+        /// <inheritdoc />
+        public override bool CanRead => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => _stream.CanWrite;
+
+        /// <inheritdoc />
+        public override bool CanTimeout => _stream.CanTimeout;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            var cipher = ReadCipher;
+            var n = _stream.Read(buffer, offset, count);
+            cipher.ProcessBytes(buffer, offset, n, buffer, offset);
+            return n;
+        }
+
+        /// <inheritdoc />
+        public override async Task<int> ReadAsync(byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            var cipher = ReadCipher;
+            var n = await _stream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+            cipher.ProcessBytes(buffer, offset, n, buffer, offset);
+            return n;
+        }
+
+        /// <inheritdoc />
+        public override void Flush() { _stream.Flush(); }
+
+        /// <inheritdoc />
+        public override Task FlushAsync(CancellationToken cancel) { return _stream.FlushAsync(cancel); }
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename parameter 'cancel' to 'cancellationToken' to match the base class declaration.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)